# Livro para consulta:
- https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html
- https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html
    

# 1. Importando bibliotecas <a name="import"></a>

<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# 2. Carregando o dataframe SINASC <a name="read"></a>
<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [2]:
df = pd.read_csv('datasus.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27028 entries, 0 to 27027
Data columns (total 69 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ORIGEM        27028 non-null  int64  
 1   CODESTAB      26913 non-null  float64
 2   CODMUNNASC    27028 non-null  int64  
 3   LOCNASC       27028 non-null  int64  
 4   IDADEMAE      27028 non-null  int64  
 5   ESTCIVMAE     26711 non-null  float64
 6   ESCMAE        26716 non-null  object 
 7   CODOCUPMAE    24121 non-null  float64
 8   QTDFILVIVO    25455 non-null  float64
 9   QTDFILMORT    24930 non-null  float64
 10  CODMUNRES     27028 non-null  int64  
 11  GESTACAO      25796 non-null  object 
 12  GRAVIDEZ      26949 non-null  object 
 13  PARTO         26979 non-null  object 
 14  CONSULTAS     27028 non-null  int64  
 15  DTNASC        27028 non-null  object 
 16  HORANASC      27007 non-null  float64
 17  SEXO          27024 non-null  object 
 18  APGAR1        26932 non-nu

# Tarefa 1

### 1. Idade media das mães e dos pais por município (coluna munResNome)


In [3]:
# colunas 'IDADEMAE' | 'munResNome'

media_idade = df.groupby('munResNome')[['IDADEMAE', 'IDADEPAI']].mean()
media_idade

IDADEMAE   IDADEPAI
munResNome                                     
Alta Floresta D'Oeste      25.991826  29.548701
Alto Alegre dos Parecis    24.844156  29.179104
Alto Paraíso               24.960177  28.833333
Alvorada D'Oeste           25.770732  30.757282
Ariquemes                  25.607866  32.466667
Buritis                    25.559177  30.812500
Cabixi                     26.037500  34.333333
Cacaulândia                25.546667  36.333333
Cacoal                     26.890102  30.931854
Campo Novo de Rondônia     24.825301  30.500000
Candeias do Jamari         25.243323  29.894737
Castanheiras               27.281250  30.392857
Cerejeiras                 27.240506  31.652778
Chupinguaia                25.333333  29.629630
Colorado do Oeste          27.602804  34.354839
Corumbiara                 24.716535  32.470588
Costa Marques              24.412017  30.613636
Cujubim                    24.482927  31.400000
Espigão D'Oeste            26.093333  31.198113
Governador Jorge Teixeira  24.401786  29.222222
Guajará-Mirim              25.462228  30.470588
Itapuã do Oeste            26.064220  42.500000
Jaru                       25.982974  32.197183
Ji-Paraná                  26.361137  31.132239
Machadinho D'Oeste         25.480000  31.326923
Ministro Andreazza         27.904762  33.849315
Mirante da Serra           26.304878  32.154762
Monte Negro                25.648241  34.678571
Município ignorado - RO    24.000000  22.000000
Nova Brasilândia D'Oeste   26.222222  30.520833
Nova Mamoré                25.326389  31.692308
Nova União                 26.894231  31.371429
Novo Horizonte do Oeste    26.938776  31.588235
Ouro Preto do Oeste        26.375594  31.233244
Parecis                    26.454545  32.473684
Pimenta Bueno              25.199675  29.745174
Pimenteiras do Oeste       25.075000  35.600000
Porto Velho                26.312670  32.352679
Presidente Médici          25.796053  30.150754
Primavera de Rondônia      27.023256  31.888889
Rio Crespo                 26.660000  28.000000
Rolim de Moura             26.600863  30.360479
Santa Luzia D'Oeste        26.121212  30.703297
Seringueiras               24.910714  30.500000
São Felipe D'Oeste         26.481481  31.395833
São Francisco do Guaporé   25.660066  31.638889
São Miguel do Guaporé      25.926346  32.666667
Teixeirópolis              26.031250  31.500000
Theobroma                  25.536082  30.083333
Urupá                      26.688235  31.008929
Vale do Anari              26.096774  34.000000
Vale do Paraíso            25.679012  29.976744
Vilhena                    26.433962  34.152941

### Logo acima temos a idade média das mães e pais de recem nascidos, separados por municício:

> Ex.: Idade média das mães na cidade de Alta Floresta D'Oeste é aproximadamente 26 anos.

### 2. Peso médio dos bebes por sexo que nasceram no dia do seu aniversário por faixas de escolaridade mae
Ex: Você, aluna(o), nasceu no dia 10/01, então você precisa filtrar o conjunto de dados nessa data e calcular o peso médio dos bebês de cada sexo por faixa de escolaridade da mãe.

In [4]:
media_peso = df[df['DTNASC'] == '2019-06-07'].groupby(
    ['ESCMAE', 'SEXO'], as_index=False)['PESO'].mean()

media_peso

ESCMAE       SEXO         PESO
0  12 anos ou mais   Feminino  3360.571429
1  12 anos ou mais  Masculino  3463.750000
2       4 a 7 anos   Feminino  3410.000000
3       4 a 7 anos  Masculino  2869.250000
4      8 a 11 anos   Feminino  3044.666667
5      8 a 11 anos  Masculino  3278.800000
6          Nenhuma   Feminino  3845.000000
7          Nenhuma  Masculino   350.000000

### No meu caso, sou nascido no dia 07 de junho, logo o dataframe está ajustado para mostrar apenas crianças que nasceram nesse mesmo dia, separados  por sexo e anos de estudo da mãe:

> Ex.: mães que tem de 4 a 7 anos de estudo, tiveram filhos do sexo masculino com peso em média de 2,86 Kg.

### 3. Qual o municipio que nasceu menos bebe em 2019?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

In [98]:

menosbbs = df.groupby('munResNome').agg(
    nascimentos = ('munResNome', 'size'),
    idade_mae_media = ('IDADEMAE', 'mean'),
    idade_mae_max = ('IDADEMAE', 'max'),
    idade_mae_min = ('IDADEMAE', 'min'),
    idade_pai_media = ('IDADEPAI', 'mean'),
    idade_pai_max = ('IDADEPAI', 'max'),
    idade_pai_min = ('IDADEPAI', 'min')
)

menosbbs_ordem = menosbbs.sort_values(by = 'nascimentos', ascending= True)
menosbbs_ordem

nascimentos  idade_mae_media  idade_mae_max  \
munResNome                                                               
Município ignorado - RO              1        24.000000             24   
Castanheiras                        32        27.281250             39   
Pimenteiras do Oeste                40        25.075000             40   
Primavera de Rondônia               43        27.023256             39   
Parecis                             44        26.454545             41   
Rio Crespo                          50        26.660000             39   
São Felipe D'Oeste                  54        26.481481             41   
Teixeirópolis                       64        26.031250             39   
Cacaulândia                         75        25.546667             42   
Cabixi                              80        26.037500             39   
Vale do Paraíso                     81        25.679012             52   
Ministro Andreazza                  84        27.904762             44   
Vale do Anari                       93        26.096774             44   
Theobroma                           97        25.536082             42   
Novo Horizonte do Oeste             98        26.938776             41   
Santa Luzia D'Oeste                 99        26.121212             41   
Nova União                         104        26.894231             39   
Itapuã do Oeste                    109        26.064220             46   
Governador Jorge Teixeira          112        24.401786             41   
Corumbiara                         127        24.716535             43   
Alto Alegre dos Parecis            154        24.844156             41   
Chupinguaia                        162        25.333333             42   
Mirante da Serra                   164        26.304878             46   
Campo Novo de Rondônia             166        24.825301             40   
Seringueiras                       168        24.910714             42   
Urupá                              170        26.688235             43   
Monte Negro                        199        25.648241             40   
Cujubim                            205        24.482927             42   
Alvorada D'Oeste                   205        25.770732             44   
Colorado do Oeste                  214        27.602804             47   
Alto Paraíso                       226        24.960177             42   
Costa Marques                      233        24.412017             41   
Cerejeiras                         237        27.240506             53   
Nova Brasilândia D'Oeste           261        26.222222             51   
São Francisco do Guaporé           303        25.660066             41   
Presidente Médici                  304        25.796053             44   
Candeias do Jamari                 337        25.243323             45   
São Miguel do Guaporé              353        25.926346             42   
Alta Floresta D'Oeste              367        25.991826             53   
Nova Mamoré                        432        25.326389             46   
Espigão D'Oeste                    450        26.093333             44   
Machadinho D'Oeste                 500        25.480000             44   
Buritis                            583        25.559177             44   
Pimenta Bueno                      616        25.199675             47   
Ouro Preto do Oeste                631        26.375594             44   
Guajará-Mirim                      781        25.462228             43   
Jaru                               881        25.982974             45   
Rolim de Moura                     927        26.600863             44   
Cacoal                            1374        26.890102             47   
Vilhena                           1590        26.433962             46   
Ariquemes                         1729        25.607866             44   
Ji-Paraná                         2182        26.361137             44   
Porto Velho                    

### Utilizando groupby do pandas, podemos observar que o município que teve o menor número de nascidos em um dia, foi o Município ignorado - RO, que contou com um nascimento.

> Nesse dia, já que tinha apenas uma mãe, a idade média, mínima e máxima das mães foi de 24 anos, já do pai como só teve apenas um, as idade média, mínima e máxima dos pais, foi de 22 anos.


### 4. Qual o municipio que nasceu mais bebe no mês de março?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

In [12]:
df['DTNASC'] = pd.to_datetime(df['DTNASC'])
marco_df = df[df['DTNASC'].dt.month == 3]

maisbbs = marco_df.groupby('munResNome').agg(
    nascimentos = ('munResNome', 'size'),
    idade_mae_media = ('IDADEMAE', 'mean'),
    idade_mae_max = ('IDADEMAE', 'max'),
    idade_mae_min = ('IDADEMAE', 'min'),
    idade_pai_media = ('IDADEPAI', 'mean'),
    idade_pai_max = ('IDADEPAI', 'max'),
    idade_pai_min = ('IDADEPAI', 'min')
)

maisbbs_ordem = maisbbs.sort_values(by = 'nascimentos', ascending= False)
maisbbs_ordem

nascimentos  idade_mae_media  idade_mae_max  \
munResNome                                                               
Porto Velho                        744        26.596774             44   
Ji-Paraná                          188        25.377660             44   
Vilhena                            148        26.412162             39   
Ariquemes                          141        25.304965             42   
Cacoal                             133        26.466165             40   
Jaru                                80        26.550000             40   
Guajará-Mirim                       75        26.333333             43   
Rolim de Moura                      69        27.594203             41   
Ouro Preto do Oeste                 67        25.522388             44   
Pimenta Bueno                       62        25.177419             39   
Buritis                             59        25.406780             42   
Nova Mamoré                         53        24.207547             42   
Espigão D'Oeste                     44        25.409091             43   
Machadinho D'Oeste                  44        26.295455             43   
São Miguel do Guaporé               41        26.365854             42   
Candeias do Jamari                  33        24.363636             40   
São Francisco do Guaporé            31        25.677419             35   
Alta Floresta D'Oeste               31        26.193548             35   
Presidente Médici                   29        25.344828             38   
Costa Marques                       24        25.666667             36   
Alvorada D'Oeste                    22        25.772727             36   
Alto Paraíso                        22        24.818182             41   
Monte Negro                         20        25.650000             36   
Nova Brasilândia D'Oeste            20        26.650000             35   
Colorado do Oeste                   19        27.736842             39   
Seringueiras                        19        25.473684             39   
Chupinguaia                         18        26.277778             42   
Cujubim                             17        24.882353             40   
Cerejeiras                          16        27.062500             33   
Urupá                               14        26.357143             38   
Vale do Paraíso                     13        23.615385             33   
Campo Novo de Rondônia              13        22.153846             35   
Corumbiara                          12        24.083333             37   
Vale do Anari                       11        24.545455             34   
Alto Alegre dos Parecis             10        25.100000             36   
Governador Jorge Teixeira           10        24.500000             33   
Cacaulândia                         10        26.200000             32   
Rio Crespo                           9        28.888889             37   
Mirante da Serra                     9        24.888889             37   
Theobroma                            8        26.875000             35   
Itapuã do Oeste                      8        25.625000             39   
Santa Luzia D'Oeste                  7        23.285714             27   
Nova União                           7        23.428571             38   
Cabixi                               6        28.000000             36   
Parecis                              6        28.000000             37   
Ministro Andreazza                   6        27.166667             35   
Pimenteiras do Oeste                 6        23.000000             29   
Primavera de Rondônia                5        26.800000             33   
São Felipe D'Oeste                   5        23.600000             31   
Novo Horizonte do Oeste              5        29.000000             37   
Castanheiras                         4        30.500000             35   
Teixeirópolis                        3        26.666667             32   

                           ida

### Utilizando groupby do pandas, podemos observar que o município que teve o maior número de nascidos em um dia, foi o município de Porto Velho, que contou com 744 nascimentos.

> Nesse dia a idade média, mínima e máxima das mães foram respectivamente de 27, 44 e 14 anos, já para os pais, a idade média, mínima e máxima dos pais, foram de respectivamente 35, 62 e 19 anos.

### Analise as respostas encontradas, tire algum insight delas, conte pra gente algo encontrado nos dados. Algo que você julgue relevante e novo pra você.

Exemplo:
- Ah, descobri que a idade mediana das mulheres que deram a luz no ano de 2019 dos municipios x é maior que y.

### Como o número de consultas pré-natal, impacta na saúde do bebê nos primeiros minutos do seu nascimento. Existe alguma correlação?

In [33]:
# Filtrando o dataframe para as colunas que serão usadas:

df1 = df[['APGAR1', 'APGAR5', 'CONSPRENAT']].dropna().reset_index()
df1 = df1.drop(columns=['index'])

In [40]:
# correlação entre a quantidade de consultas pré-natal e APGAR1:

corr_con_apgar1 = df1['CONSPRENAT'].corr(df1['APGAR1'])
corr_con_apgar1

-0.006317218193724421

In [41]:
# correlação entre a quantidade de consultas pré-natal e APGAR5:

corr_con_apgar5 = df1['CONSPRENAT'].corr(df1['APGAR5'])
corr_con_apgar5

-0.0038707509612093045

#### Não existe nenhuma relação linear:

> Através dos resultados das correlações, podemos observar que por incrivel que pareça, não existe nenhuma relação entre o número de **consultas pré-natal** e o grau de **APGAR** do bebê.

- Vale resaltar que essa resposta se restringe a essa base de dados